dvc/remote/ssh.py

```diff
@@ -392,3 +388,29 @@ def list_cache_paths(self):
         flist = stdout.split()
         ssh.close()
         return flist
+
+    def should_ask_for_password(self):
+        from paramiko.dsskey import DSSKey
+        from paramiko.ecdsakey import ECDSAKey
+        from paramiko.ed25519key import Ed25519Key
+        from paramiko.rsakey import RSAKey
+        from paramiko.ssh_exception import (
+            SSHException,
+            PasswordRequiredException,
+        )
+
+        if self.password:
+            return False
+
+        if not self.keyfile:
```

Comment:
And what if auth is password-only and it is not specified? E.g. imagine you logging into your ssh server by only entering a password(don't confuse it with a passphrase for a private key). Seems like we should leave `ask_password` for that scenario, right?

Reply:
@efiop , could you write a small example of the imaginary scenario? (i.e. `dvc remote modify`...)

Reply:
@efiop , by password-only approach you mean to have the following config on the server:
```
AuthenticationMethods=password
```

Reply:
```
dvc remote add ssh://127.0.0.1:/path/to/dir
dvc push
```

Just plain simple no-private-keys auth, like `ssh example.com -> Enter a password for example.com`.

Reply:
@mroutis are you able to `ssh 127.0.0.1` without a password?

Reply:
I'm using a container to tesh `ssh` on my localhost (currently there's no running ssh daemon on my machine) :stuck_out_tongue: 

But I got you, when there's no password to use, the user will need to set `dvc remote modify ssh ask_password false`.

Do you think the PR is worth it, tho? Implying that the normal use case wouldn't be a server with `PermitEmptyPasswords` (I hope :stuck_out_tongue: )

Reply:
@mroutis I don't mean an empty password(though it is actually an interesting case as well). I meant a proper password auth. In the example `dvc remote + dvc push` above, user doesn't have a private key configured, so when he does `ssh example.com` ssh asks him for a password to enter, right? In our case we won't detect that(i don't think if we even care), so the only option for a user would be to set `dvc remote modify myssh password MYPASS`, which is not secure. With `ask_password true` he could ask dvc to prompt him for a password at the runtine(which is how this works right now). Your PR is still extremely useful for cases when user is using a private key that is protected by a passphrase.

Reply:
:thinking: , @efiop , but the following would ask for a password when there's no `keyfile` configured:
```python
        if not self.keyfile:
            return True
```

Reply:
Ok, and now imagine you have generated a keyfile access without a password(key is placed in ~/.ssh, and you don't specify it with `dvc remote modify`). So when you do `ssh example.com` it doesn't ask you for a password at all. How would dvc work in that scenario?

Reply:
Got it, so we need to check for the existance of `~/.ssh/id_dsa`, `~/.ssh/id_ecdsa`, `~/.ssh/id_ed25519` and `~/.ssh/id_rsa`, right?

Reply:
Also, we need to check if there's any key on the SSH agent, as the user may have provided the password before (and will break the whole idea of having an agent).

Reply:
Dang, there's a lot of edge cases that paramiko already handles on their `connect` method. Maybe it is better to wrap the `ssh.connect` method into a `try/except PasswordRequiredException` then, we prompt for the password. I'll try that

Reply:
> Got it, so we need to check for the existance of `~/.ssh/id_dsa`, `~/.ssh/id_ecdsa`, `~/.ssh/id_ed25519` and `~/.ssh/id_rsa`, right?

@mroutis Seems like a lot of hastle :) Let's just remove
```
   if not self.keyfile:
            return True
```
and we will be able to implicitly cover explicitly configured keyfiles with a password and will still be able to explicitly handle the cases we've discussed above. This is nice and simple. What do you think?

Reply:
Or yes, maybe `try/except PasswordRequiredException` would work.

Reply:
@efiop , `try/except` works but it breaks on the unit tests, something about `socket.gaierror`, I'm not sure why it is :confused: .

I'm going to move to another issue; I thought that it wouldn't take me that much but we need to take into account a lot of stuff that is better to delegate it to `paramiko`, the problem with that is issuing a `ssh.connect -> ssh.close` just for the sake of checking if we need to ask for a password or not.

Removing `if not self.keyfile` would break the first test case:
```
        config = {"url": "ssh://127.0.0.1/tmp"}
        remote = RemoteSSH(None, config)
        self.assertTrue(remote.should_ask_for_password())
```

What I think is happening on the case of the case you mentioned:
- `~/.ssh/id_rsa` password-less key
- `dvc init && dvc remote add ssh ssh://localhost/tmp`
- `ssh localhost` (wouldn't ask for the password)

Is that the `ssh` client is searching for default identity files (the ones that I listed above, including `~/.ssh/id_rsa`) and then it checks if the keyfile is encrypted or not (by raising the `PasswordRequiredException`).

Maybe I'll come back later to this one, but feel free to send a commit if you think is already there or have another solution  :)

Thanks for your review / back-and-forth, very useful :+1: !

Reply:
@mroutis 

That test case is not right. Password should be asked only if `ssh 127.0.0.1` asks for a password, which we are unable to know in advance right now, so that is why we rely on `ask_password` config.

Let's not leave this one hanging. At least format this PR to fix progress bar interruption, so we could merge it. If you think that this password detection is too much, then just note that under the issue and close it, no need to just keep it around if it is too much hustle and no one is asking for it. We already have explicit `ask_password` which covers all the cases, even though it is not as nice as automatic detection.

Reply:
Totally, @efiop , I'll close it and fix the progress bar interruption in another PR.